File: Bit Manipulation/Medium/1310. XOR Queries of a Subarray/1310. XOR Queries of a Subarray.c
#include <stdlib.h>

/**
 * Problem: XOR Queries of a Subarray
 * Description: Given an array `arr` and a list of queries, where each query is represented by two integers `x` and `y`, 
 * you are to return the XOR of the elements in the subarray between indices `x` and `y` (inclusive).
 *
 * Note: The returned array must be malloced, assume the caller calls free().
 *
 * @param arr: The input array of integers.
 * @param arrSize: The size of the input array.
 * @param queries: The 2D array of queries where each query is represented by two integers.
 * @param queriesSize: The number of queries.
 * @param queriesColSize: The column sizes for the queries (2 for all queries).
 * @param returnSize: Pointer to store the size of the returned result array.
 * @return: An array of integers where each value is the result of a query.
 */
int* xorQueries(int* arr, int arrSize, int** queries, int queriesSize, int* queriesColSize, int* returnSize) {
    // Compute the prefix XOR array in-place
    for (int i = 1; i < arrSize; i++) {
        arr[i] = arr[i - 1] ^ arr[i];
    }

    // Allocate memory for the result array
    int *ans = (int*)calloc(queriesSize, sizeof(int));

    // Process each query
    for (int j = 0; j < queriesSize; j++) {
        int x = queries[j][0];
        int y = queries[j][1];
        if (x > 0) {
            ans[j] = arr[y] ^ arr[x - 1];
        } else {
            ans[j] = arr[y];
        }
    }

    // Set the returnSize to the number of queries
    *returnSize = queriesSize;

    return ans;
}

## Time Complexity: O(N + Q), where N is the size of the array and Q is the number of queries.
## Space Complexity: O(Q), where Q is the size of the result array.

File: Bit Manipulation/Medium/1310. XOR Queries of a Subarray/1310. XOR Queries of a Subarray.py
# Problem: XOR Queries of a Subarray
# Description: You are given an array `arr` and a list of queries `queries` where each query is represented as a list `[x, y]`. For each query, the task is to find the XOR of elements from index `x` to `y` in the array `arr`.

from typing import List

class Solution:
    def xorQueries(self, arr: List[int], queries: List[List[int]]) -> List[int]:
        """
        Given an array 'arr' and a list of queries, this function returns the XOR of elements 
        between each query range [x, y].

        :param arr: List of integers.
        :param queries: List of queries where each query is a list of two integers [x, y].
        :return: List of results, where each result is the XOR of elements between indices x and y.
        """
        n = len(arr)
        # Create an auxiliary array where xor[i] stores the XOR from 0 to i-1
        xor = [0] * (n + 1)
        
        # Fill the XOR array
        for i in range(n):
            xor[i + 1] = xor[i] ^ arr[i]
        
        ans = []
        
        # Process each query
        for query in queries:
            x = query[0]
            y = query[1]
            # The XOR for range [x, y] is computed as xor[y+1] ^ xor[x]
            ans.append(xor[x] ^ xor[y + 1])
        
        return ans

# Time Complexity: O(n + m), where n is the length of the array `arr` and m is the number of queries.
# The preprocessing step takes O(n), and each query is processed in constant time O(1).
# Space Complexity: O(n), due to the extra space needed for the `xor` array.

# Example usage:
# solution = Solution()
# arr = [1, 3, 4, 8]
# queries = [[0, 1], [1, 2], [0, 3], [3, 3]]
# print(solution.xorQueries(arr, queries))  # Expected output: [2, 7, 14, 8]
